akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -753,27 +779,31 @@ private[akka] class DDataShard(
     stateKeys(i)
   }
 
-  // get initial state from ddata replicator
-  getState()
+  val initializedPromise = Promise[Done]
```

Comment:
not needed?

Reply:
It is used as the return of onLeaseAcquired() which returns a Future that once it is complete the ShardInitialized is sent to the parent. Maybe it is more clear for the `DData`/`Persistent` shard to just do that rather than rely on a mechanism in `Shard`

Reply:
I think mixing futures and actors like this is very hard to follow and error prone so it would be preferred if we can do it with plain actor things. I can take a look again but I didn't see why it couldn't be plain strict method calls.

Reply:
I tried some refactoring and here is my proposal: 96795138e17c09ecf16ecb62e74375858b715ae1

We have 3 cases and a summary of the flows looks like this:

* Shard
  * preStart
    * acquireLeaseIfNeeded
      * awaitingLease
  * receive LeaseAcquireResult in awaitingLease
    * onLeaseAcquired
      * parent ! ShardInitialized
      * become receiveCommand

* PersistentShard
  * receive RecoveryCompleted
    * acquireLeaseIfNeeded
      * awaitingLease
  * receive LeaseAcquireResult in awaitingLease
    * onLeaseAcquired
      * parent ! ShardInitialized
      * become receiveCommand
      * restartRememberedEntities
      * unstashAll

* DDataShard
  * preStart
    * acquireLeaseIfNeeded
      * awaitingLease
  * receive LeaseAcquireResult in awaitingLease
    * onLeaseAcquired
      * getState
      * become waitingForState
  * recieve GetSuccess in waitingForState
    * recoveryCompleted
      * parent ! ShardInitialized
      * become receiveCommand
      * restartRememberedEntities
      * unstashAll

Reply:
Yep i prefer each extension of shard doing the parent ! ShardInitialized rather than trying to do it in the base class. :+1: